src/ir/possible-contents.cpp

```diff
@@ -0,0 +1,1644 @@
+
+      // TODO: if the old contents here were an exact type then we have an old
+      //       link here that we could remove as it is redundant (the cone
+      //       contains the exact type among the others). But removing links
+      //       is not efficient, so maybe not worth it.
+    }
+
+    // Link to the canonical location.
+    connectDuringFlow(UniqueLocation{coneReadIndex},
+                      ExpressionLocation{read, 0});
+  }
+}
+
+void Flower::writeToData(Expression* ref, Expression* value, Index fieldIndex) {
+#if defined(POSSIBLE_CONTENTS_DEBUG) && POSSIBLE_CONTENTS_DEBUG >= 2
+  std::cout << "    add special writes\n";
+#endif
+
+  // We could set up links here as we do for reads, but as we get to this code
+  // in any case, we can just flow the values forward directly. This avoids
+  // increasing the size of the graph.
+  //
+  // Figure out what to send in a simple way that does not even check whether
+  // the ref or the value was just updated: simply figure out the values being
+  // written in the current state (which is after the current update) and
+  // forward them.
+  auto refContents = getContents(getIndex(ExpressionLocation{ref, 0}));
+  auto valueContents = getContents(getIndex(ExpressionLocation{value, 0}));
+  if (refContents.isNone() || refContents.isNull()) {
+    return;
+  }
+  if (refContents.hasExactType()) {
+    // Update the one possible type here.
+    auto heapLoc =
+      DataLocation{refContents.getType().getHeapType(), fieldIndex};
+    updateContents(heapLoc, valueContents);
+  } else {
+    assert(refContents.isMany() || refContents.isGlobal());
+
+    // Update all possible subtypes here.
+    auto type = ref->type.getHeapType();
+    for (auto subType : subTypes->getAllSubTypes(type)) {
+      auto heapLoc = DataLocation{subType, fieldIndex};
+      updateContents(heapLoc, valueContents);
+    }
+  }
+}
+
+void Flower::flowRefCast(const PossibleContents& contents, RefCast* cast) {
+  // RefCast only allows valid values to go through: nulls and things of the
+  // cast type. Filter anything else out.
+  PossibleContents filtered;
+  if (contents.isMany()) {
+    // Just pass the Many through.
+    // TODO: we could emit a cone type here when we get one, instead of
+    //       emitting a Many in any of these code paths
+    filtered = contents;
+  } else {
+    auto intendedType = cast->getIntendedType();
+    bool mayBeSubType =
+      HeapType::isSubType(contents.getType().getHeapType(), intendedType);
+    if (mayBeSubType) {
+      // The contents are not Many, but they may be a subtype of the intended
+      // type, so we'll pass them through.
+      // TODO: When we get cone types, we could filter the cone here.
+      filtered.combine(contents);
+    }
+    bool mayBeNull = contents.getType().isNullable();
+    if (mayBeNull) {
+      // A null is possible, so pass that along.
+      filtered.combine(
+        PossibleContents::literal(Literal::makeNull(intendedType)));
+    }
```

Comment:
There are other places we call `combine`. Why do we do this null check only here?

Reply:
Normally we just call `combine()` on all the arriving contents. The tricky thing here is that `ref.cast` filters some things out. So we basically check for the things it would allow through, and pass them through (and so nothing else gets by). Those things are either a proper subtype, or a null, so we have logic for both of those things here.